Vxsort

PeterSolMS opened the pull request:
Faster sorting code from Dan Shechter, and bigger mark list.

Not ready to be merged yet, but ready for initial review.

PeterSolMS committed:
Initial snapshort of vxsort implementation incorporating source code from Dan Shechter.

PeterSolMS committed:
Bug fix from Dan.

PeterSolMS committed:
Use bigger mark list for experiments.

PeterSolMS committed:
Give up if the mark list size is bigger than a reasonable fraction of the ephemeral space.

stephentoub commented:
cc: @damageboy

jkotas commented on src/coreclr/src/gc/vxsort.cpp:
```diff
@@ -0,0 +1,282 @@
+
```
License header?

It may need an entry in
https://github.com/dotnet/runtime/blob/master/THIRD-PARTY-NOTICES.TXT

damageboy replied:
I don't mind adding whatever license you feel is friendliest to the original

damageboy replied:
In the meantime, I've licensed under MIT license:
https://github.com/damageboy/gcsort/blob/master/LICENSE

jkotas commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2060,6 +2068,42 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
+    };
+}
+void vxsort(uint8_t** low, uint8_t** high, unsigned int depth)
+{
```
Do we need "is avx available" check here and fallback to the non-vectorized sort ?

damageboy replied:
Yeap, unless you'd like me/Peter to copy the rest of the non-vectorized partitioning back into the same sort.

It already shares the heap-sort + insertion-sort code from the original as a fallback just in case we run into one of the existing edge-cases, as with standard introsort.

This should reduce duplication, as I literally copied 2/3 of introsort into vxsort from the get go.

damageboy replied:
With clang/gcc I've previously used [builtins](https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html):
```c 
if (__builtin_cpu_supports("avx2")) {
  // AVX2
} else {
  // SAD PANDA
}
```

To cleanly detect this.

Pretty sure you'll have to do something else for MSVC.

PeterSolMS replied:
Do we have prior art for this? I found EEJitManager::SetCpuInfo which detects the AVX2 instruction set for the JIT.

jkotas replied:
For Unix, `__builtin_cpu_supports("avx2")` looks pretty nice. For Windows, I guess it will need to call `GetEnabledXStateFeatures` and check for `XSTATE_MASK_AVX`

damageboy replied:
The general mechanism I'm familiar with is using function multi-versioning, this is supported in gcc since forever, and in clang from 7.0 onwards if I'm not mistaken.

The basic idea would be that a function can be decorated with:
```c
__attribute__ ((target ("default")))
int foo () {

  return 1;
}
__attribute__ ((target ("avx2")))
int foo () {
  return 2;
}
```

And you get an indirect jump embedded in the code stream courtesy of the compiler.

I'm pretty sure this does not exist for MSVC, where you need to hand-code such things, but then again, you have more apt MSVC gurus around you, so they might know something I don't. But it seems that if you would like to avoid depending on the EEJitManager, I could add support for this myself for clang/gcc the normal mechanisms, and for MSVC by following what seems to be the [standard practice](https://docs.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?redirectedfrom=MSDN&view=vs-2019) inside vxsort, that way I can switch from AVX2, AVX512 when my AVX512 support is complete.

Would you like me to take care of all of this internally and expose one sort functionality that does everything?

tannergooding replied:
Is there any reason we can't use the cached checks the VM already has from the HWIntrinsics feature?

jkotas replied:
We want to support building GC into separate .dll (aka LocalGC).

damageboy replied:
I'll also add that you don't detect AVX512 yet either, which I'm working on in a vxsort branch, regardless if @PeterSolMS ends up taking it.

jkotas commented on src/coreclr/src/gc/bitonic_sort.int64_t.generated.h:
```diff
@@ -0,0 +1,1481 @@
+
```
Comment about how this file was generated would be nice

damageboy replied:
The script is part of my repo:

https://github.com/damageboy/gcsort/blob/master/gcsort/smallsort/bitonic_gen.py

Note that it generates all 6 32+64 bit types.

damageboy replied:
By the way, I would highly recommend including the generator and reviewing it instead of the mountains of code it generated.

It might actually make sense that way :)

damageboy replied:
@jkotas I've updated the [original repo](https://github.com/damageboy/gcsort/blob/master/gcsort/smallsort/bitonic_sort.int64_t.generated.h) to now include proper auto-generation disclaimers:

```c
/////////////////////////////////////////////////////////////////////////////
////
// This file was auto-generated by a tool at 2020-05-30 11:18:44
//
// It is recommended you DO NOT directly edit this file but instead edit
// the code-generator that generated this source file instead.
/////////////////////////////////////////////////////////////////////////////
```

I'm assuming we'll do another pass of updating the gc copy from my original given there will be more changes stemming from this review process.

Maoni0 replied:
we'd definitely want to include the generator. 
BTW @PeterSolMS is OOF, that's why he hasn't said anything, he will be back this coming Tue.

PeterSolMS replied:
I wonder whether to put the generator under src/tools or src/gc?

jkotas replied:
We typically place generators like this locally to the thing they are generating. Here are a few existing examples:
https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/CrstTypeTool.cs
https://github.com/dotnet/runtime/tree/master/src/coreclr/src/System.Private.CoreLib/Tools/GenUnicodeProp

damageboy commented:
For what it's worth, I've done some initial work to clean up the original codebase.

Currently the gcsort repo's latest and greatest can build on Mac, Linux, and windows, on a combination of msvc+clang-cl for windows and clang+gcc for mac+linux.

There is no functional change to the c++ code in the original, but knowing that many compilers and operating system all [build and run tests](https://github.com/damageboy/gcsort/actions/runs/120882529) successfully should provide some reassurance as to the quality of the code.

The macOs build, specifically, cannot run tests on GH Actions right now, as it seems those machines lack AVX2 support? which seems odd enough, but still...

.

PeterSolMS committed:
Latest version from Dan, disable for ARM64.

PeterSolMS committed:
Fixes for Linux compile, ended up disabling vxsort for Linux for now.

PeterSolMS committed:
Experimenting with 32-bit sort - the variation that gathers mark list entries pertaining to the local heap by reading the mark lists from all the heaps appears to be too slow and scales very badly with increasing number of heaps.

PeterSolMS committed:
32-bit sort - preserve failing case.

danmosemsft commented:
For the sake of us curious bystanders, do we have an idea of the impact of this cool change? Eg., perf measurements.

Maoni0 commented:
measuring and analyzing perf impact is a big part of what Peter has been doing; and will be included as part of the commit message when we merge.

danmosemsft commented:
Sounds good - I was just excited :)

PeterSolMS committed:
Do the pointer compression/decompression in place, to improve performance, optionally write mark lists and associated information to binary files for further analysis.

PeterSolMS committed:
Merge branch 'master' into vxsort

PeterSolMS committed:
Introduce runtime check whether CPU supports AVX2 instruction set.

PeterSolMS committed:
Implement mark list growth.

PeterSolMS committed:
Integrate new version including AVX512 code path.

PeterSolMS committed:
Implement runtime test for AVX512 support.

PeterSolMS committed:
Move the files for the vectorized sort to their own directory, add stubs to call AVX2 or AVX512 flavor of the sort.

PeterSolMS committed:
Get rid of unneeded #include statement in two files.

tannergooding commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2084,46 +2084,47 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
-
-    template <>
-    class numeric_limits<int64_t>
+        do_vxsort_avx512(low, high);
```
Are there any numbers showing how AVX512 performs in real world scenarios?

It is fairly well documented that using AVX512 *can* cause downclocking which in certain circumstances can lead to decreased overall application performance (AVX2 can as well, but to a much lesser extent).
* The impact is lower on newer processors such as `Ice Lake` as compared to older `Skylake Server`
* The impact is typically relegated to instructions that use the floating-point or multiplication (including FMA) pipelines
* The impact is typically for 256 and 512-bit variants of the instructions, rather than 128-bit

The official guidance from Intel is the following (from the May 2020 edition of [Intel 64 and IA-32 Architectures Optimization Reference Manual](https://software.intel.com/content/www/us/en/develop/download/intel-64-and-ia-32-architectures-optimization-reference-manual.html)):
![image](https://user-images.githubusercontent.com/10487869/85576402-3b994b00-b5ed-11ea-81ab-5b02ac5e5ac1.png)

The section detailing `Skylake Server Power Management` is a bit longer and includes more details on what impacts each frequency level and includes various graphs/examples of the typical impact, with the explicit guidance to still profile.

damageboy replied:
Downclocking can be pretty severe on 14nm parts right now:
From Anandtech's article [Sizing Up Servers: Intel's Skylake-SP Xeon versus AMD's EPYC 7000 - The Server CPU Battle of the Decade?](https://www.anandtech.com/show/11544/intel-skylake-ep-vs-amd-epyc-7000-cpu-battle-of-the-decade/8)
![image](https://user-images.githubusercontent.com/125730/85596303-4639fb80-b652-11ea-8eb7-b7cf96683dba.png)

While that is true, we are still seeing substantial perf boost on a given machine (copy pasting some preliminary results, while I'm still improving the code behind the scenes):

For 64-bit sorting, here is the original + 8way unrolled results, obtained from a shared instance running a Xeon Silver 4216 with nominal speed @ 2.1Ghz:

```
---------------------------------------------------------------------------------------------------
Benchmark (<type/vector-isa/unroll>/size/threads)                      Time/N (per Element)
---------------------------------------------------------------------------------------------------
BM_full_introsort/65536/threads:1                                      56.55590 ns
BM_full_introsort/131072/threads:1                                     60.63450 ns
BM_full_introsort/262144/threads:1                                     65.21260 ns
BM_full_introsort/524288/threads:1                                     69.03150 ns
BM_full_introsort/1048576/threads:1                                    73.16970 ns

BM_vxsort<int64_t, vector_machine::AVX2, 8>/65536/threads:1            20.14540 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/131072/threads:1           21.08380 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/262144/threads:1           22.36290 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/524288/threads:1           23.05820 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/1048576/threads:1          24.21020 ns

BM_vxsort<int64_t, vector_machine::AVX512, 8>/131072/threads:1          8.23500 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/262144/threads:1          8.90877 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/524288/threads:1          9.94376 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/1048576/threads:1        10.70730 ns
```

So it's pretty clear the AVX512 is out-performing AVX2, post-downclocking.
There is also a very clear reason for this staggering improvement, if you are aware of two key points:
* AVX2 is missing some int64 functionality (certain int64 ops are more expensive)
* There is an extra specific AVX512 intrinsic (_mm512_compress_storeu_epiXX), which is removing the entire lookup table + cache reference that was involved for loading the lookup entry (which is a huge win)

As such, the perf bump can be thought of as *ONLY* being 2x-ish due to downclocking, with a clear expectation of seeing 3x with future 10nm parts and below.

As for int32 (which is also probably going to be part of this PR, given that we will probably dynamically switch to using int32 for smaller mark-lists):

```
---------------------------------------------------------------------------------------------------
Benchmark (<type/vector-isa/unroll>/size/threads)                      Time/N (per Element)
---------------------------------------------------------------------------------------------------
BM_vxsort<int32_t, vector_machine::AVX2, 8>/65536/threads:1            5.47985 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/131072/threads:1           5.89806 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/262144/threads:1           6.13415 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/524288/threads:1           6.49738 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/1048576/threads:1          7.10343 ns
 
BM_vxsort<int32_t, vector_machine::AVX512, 8>/65536/threads:1          3.70457 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/131072/threads:1         3.94963 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/262144/threads:1         4.12517 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/524288/threads:1         4.35658 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/1048576/threads:1        4.86922 ns
```

While the perf boost from AVX2 to AVX512 is smaller for int32, it is still substantial. Also, please remember these are preliminary results that are somewhat less than optimal given the lack of direct HW I'm experiencing, and that fact that everything I do with AVX512F is simply more painful because of this.

The benchmarks and code are all up-to-date and accessible from the [vxsort-cpp](https://github.com/damageboy/vxsort-cpp)

Finally, there is also a discussion to be had about how long do the downclocking effects linger on **POST** sorting.
As to that, the long answer is that Travis Downs (@trav_downs) measured this independently: https://travisdowns.github.io/blog/2020/01/17/avxfreq1.html

The short answer is no more than ~700usec:

What | Time | Description
-- | -- | -- | 
Voltage Transition | ~8 to 20 μs | Time required for a voltage transition, depends on the frequency 
Frequency Transition | ~10 μs | Time required for the halted part of a frequency transition 
Relaxation Period | ~680 μs | Time required to go back to a lower power license, measured from the last instruction requiring the higher license 

It is important to note that on server CPUs, this normally applies for individual cores, where on client CPUs this is "global".
For 14nm there are currently only server parts, with the exception of Intel Icelake CPUs.

I think that given we are saving multiple milliseconds with this optimization, even while increasing the mark-list, and paying with 700usec of reduced frequency post that time-saving window directly is more than a reasonable and temporary sacrifice.

My conclusion from this is that the current "state of the union" is roughly the worst it can get (although multi-threaded results are missing, due to me not having access to a dedicated machine where I could produce meaningful results: I barely respect the results I have, in that sense since this is not a machine I can kick people off of).

As you mentioned, Icelake and Tigerlake make things only better, and it is hard to imagine AMD doing worse once they come around to supporting AVX512, probably sometime around transitioning to TSMC 5nm.

I still think I can kick things up with a few more iterations (e.g. coming weekends). Peter has inspired me with an idea that can truly go ballistic, and I have some more ammo left regardless. So this really is as "bad" as it gets.

Hope this helps clear the fog and justify the risk.

tannergooding replied:
The main thing I'm worried about is "micro-benchmark" vs "real-world".

Its pretty easy to define a microbenchmark that shows AVX-512 or AVX2 is faster. The concern is then how that impacts everything that runs after that is done.

For example, given the numbers above AVX-512 is 2-5x faster (depending on if compared to AVX2 or SSE). However, that is ~10ns vs ~50ns. If the downclocking can last up to 700 microseconds, that is 14000x longer than the SSE scenario: https://www.google.com/search?q=700+microseconds+%2F+50+nanoseconds and the question is whether the gains outweigh potentially running the entire processor at 50-75% speed for that much longer.
* This is naturally assuming a worst or near worst case scenario, it may not actually be that bad; I just think it is something that we should carefully measure and with real world workloads, rather than just micro-benchmarks

damageboy replied:
Well, it's 10ns vs 70ns **per element**.
Multiplied by 2<sup>20</sup> it is... more...

Agreed that the only thing that counts in the end of the day is real life perf for the end-to-end case rather than benchmarks. But the more the benchmarks end up being overwhelmingly extreme compared to the starting point, at some point even real life's ability to mess with you comes to an end.

Sorting, and in these sizes (more importantly) tends to always go to the extreme side of the spectrum for the "Optimize at all costs" arguments.

We can hopefully allow larger and larger mark-lists with this, which as far as I can tell has all kinds of all-around healing effects, though I'm not a GC guru.

PeterSolMS replied:
There is a real world motivation here - we have seen real world scenarios where mark list sorting takes several milliseconds (with ~100k mark list entries). We currently have a limitation at 150k entries - when we run out, we have to take a less efficient path in plan_phase, which runs into the tens of milliseconds for real world scenarios that I've seen.

I will consider limiting the AVX512 sorting to large mark lists though, to make sure we get a sizeable benefit.

tannergooding commented on src/coreclr/src/inc/corinfoinstructionset.h:
```diff
@@ -59,6 +59,7 @@ enum CORINFO_InstructionSet
     InstructionSet_SSE2_X64=25,
     InstructionSet_SSE41_X64=26,
     InstructionSet_SSE42_X64=27,
+    InstructionSet_AVX512=28,
```
This needs to be `AVX512F` to match the actual ISA name. There are a number of AVX512 extensions and it will be important to differentiate them if we eventually extend HWIntrinsics to support them.

damageboy replied:
Agreed, this is indeed the naming on my original repo:
MSVC:
https://github.com/damageboy/vxsort-cpp/blob/master/vxsort/isa_detection_msvc.cpp#L53

clang:
https://github.com/damageboy/vxsort-cpp/blob/master/vxsort/isa_detection_sane.cpp#L20

jkotas replied:
Would it be better for GC to have its own local detection (e.g. based on @damageboy original repo), and not could it with the JIT/EE interface and intrinsics support in the JIT?

This instruction set enum is autogenerated (look for "DO NOT EDIT THIS FILE! IT IS AUTOGENERATE" at the top of the file). Adding to this enum would mean updating the auto-generator and regenerating all places where this shows up.

tannergooding replied:
I don't have a preference here. I would think it should be named `AVX512F` even if it is a separate define/detection/etc.

PeterSolMS replied:
I totally missed the fact that the enum is autogenerated - back to the drawing board... Agree with naming it AVX512F, of course.

PeterSolMS committed:
Address codereview feedback to specifically say AVX512F instead of just AVX512 as there are multiple subsets.

PeterSolMS committed:
Fix CMakeLists.tx files for non-x64 non-Windows targets, introduce separate max mark list sizes for WKS, remove dead code from grow_mark_list, add #ifdef to AVX512 detection to make the other architectures build.

PeterSolMS committed:
Instead of modifying the tool-generated header file corinfoinstructionset.h, modify InstructionSetDesc.txt that it is generated from, and run the tools that generates all the files from it.

jkotas commented on src/coreclr/src/inc/corinfoinstructionset.h:
```diff
@@ -50,15 +50,16 @@ enum CORINFO_InstructionSet
-    InstructionSet_SSE2_X64=25,
-    InstructionSet_SSE41_X64=26,
-    InstructionSet_SSE42_X64=27,
+    InstructionSet_AVX512F=19,
```
This also needs JIT/EE interface GUID update because of is revving JIT/VM contract. It has other side-effects that are not ideal e.g. AVX512F will be recognized as instruction set on the crossgen command line, even though crossgen won't ever do anything with it.

I think it would be a lot simpler and maintainable to just do a local detection in GC. It is really only needed for Windows - we can just use the built-in function multi-versioning for Unix.

PeterSolMS replied:
Ok, I implemented local detection in the GC instead and undid the changes on the VM side.

jkotas replied:
Thanks!

PeterSolMS committed:
Move AVX2/AVX512 instruction set detection to GC side.

drieseng commented on src/coreclr/src/gc/vxsort/defs.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef VXSORT_DEFS_H
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/do_vxsort.h:
```diff
@@ -0,0 +1,13 @@
+// Enum for the GCToOSInterface::SupportsInstructionSet method
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/do_vxsort_avx2.cpp:
```diff
@@ -0,0 +1,58 @@
+#include "common.h"
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/do_vxsort_avx512.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "common.h"
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/isa_detection.cpp:
```diff
@@ -0,0 +1,134 @@
+// ISA_Detection.cpp : Diese Datei enthält die Funktion "main". Hier beginnt und endet die Ausführung des Programms.
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/isa_detection_dummy.cpp:
```diff
@@ -0,0 +1,15 @@
+// ISA_Detection.cpp : Diese Datei enthält die Funktion "main". Hier beginnt und endet die Ausführung des Programms.
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/machine_traits.avx2.cpp:
```diff
@@ -0,0 +1,287 @@
+#include "common.h"
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/machine_traits.avx2.h:
```diff
@@ -0,0 +1,243 @@
+//
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/machine_traits.avx512.h:
```diff
@@ -0,0 +1,217 @@
+//
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/machine_traits.h:
```diff
@@ -0,0 +1,36 @@
+//
```
nit: missing license header

drieseng commented on src/coreclr/src/gc/vxsort/smallsort/bitonic_sort.AVX2.int64_t.generated.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "common.h"
```
nit: missing license header (do you do this for generated files?)

PeterSolMS committed:
Use vectorized packer, switch packed range from uint32_t to int32_t, because that makes the sorting a bit more efficient.

PeterSolMS committed:
Add GCConfig setting to turn vectorized sorting off, streamline ISA detection (but require initialization), rename to IsSupportedInstructionSet.

benaadams commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2088,8 +2088,9 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
-    if (SupportsInstructionSet(InstructionSet::AVX512F))
+    assert(IsSupportedInstructionSet(InstructionSet::AVX2));
+    // use AVX512F only if the list is large enough to pay for downclocking impact
+    if (IsSupportedInstructionSet(InstructionSet::AVX512F) && ((high -low) > 128*1024))
```
The thresholds `128*1024` are used twice for AVX512F and `8*1024` for AVX2; maybe take them to consts? (or GC config values?)

PeterSolMS replied:
Done - thanks!

PeterSolMS committed:
Several small improvements:
 - Don't waste time sorting the mark list if background GC is running as we are not going to use it.
 - Use smaller max mark list size if we cannot use AVX2/AVX512 instruction sets
 - Fix mark list overflow detection for server GC.

PeterSolMS committed:
Address codereview feedback - add constants for the thresholds above which we use AVX/AVX512F instruction sets.

Add space before parameter lists as per GC codebase coding conventions.

Improve some comments.

PeterSolMS committed:
Add license headers and entry in THIRD-PARTY_NOTICES.TXT for Dan's vectorized sorting code.

jkotas commented on src/coreclr/src/gc/vxsort/alignment.h:
```diff
@@ -0,0 +1,40 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
We have standardized on two-line version of the license header in https://github.com/dotnet/runtime/pull/38793

PeterSolMS replied:
Ok, I've updated the license headers to the two-line version.

PeterSolMS committed:
Update license headers

Maoni0 commented:
still looking...

PeterSolMS commented on src/coreclr/src/gc/gcconfig.h:
```diff
@@ -129,6 +129,7 @@ class GCConfigStringHolder
     INT_CONFIG   (GCHeapHardLimitSOHPercent, "GCHeapHardLimitSOHPercent", NULL,                             0,                 "Specifies the GC heap SOH usage as a percentage of the total memory")              \
     INT_CONFIG   (GCHeapHardLimitLOHPercent, "GCHeapHardLimitLOHPercent", NULL,                             0,                 "Specifies the GC heap LOH usage as a percentage of the total memory")              \
     INT_CONFIG   (GCHeapHardLimitPOHPercent, "GCHeapHardLimitPOHPercent", NULL,                             0,                 "Specifies the GC heap POH usage as a percentage of the total memory")              \
+    INT_CONFIG   (GCEnabledInstructionSets, "GCEnabledInstructionSets", NULL,                        -1,                 "Specifies whether GC can use AVX2 or AVX512F")                                           \
```
Sure, I just added to the comment that is already part of the line:

"Specifies whether GC can use AVX2 or AVX512F - 0 for neither, 1 for AVX2, 3 for AVX512F"

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -19,7 +19,11 @@
 
 #include "gcpriv.h"
 
+#if defined(TARGET_AMD64) && defined(TARGET_WINDOWS)
```
I was working on enabling vxsort in Linux, I will add it to this PR if you don't mind.

PeterSolMS replied:
As it turns out, enabling this for Linux is more involved than I thought - I think this should go in a separate PR after all.

PeterSolMS commented on src/coreclr/src/gc/vxsort/isa_detection.cpp:
```diff
@@ -0,0 +1,140 @@
+void InitSupportedInstructionSet (int32_t configSetting)
+{
+    s_supportedISA = (SupportedISA)((int)DetermineSupportedISA() & configSetting);
+    // we are assuming that AVX2 can be used if AVX521F can,
```
Thanks for the catch!

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -2079,6 +2083,59 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
+const size_t AVX2_THRESHOLD_SIZE = 8 * 1024;
+
+// above this threshold, using AVX51F for sorting will likely pay off
+// despite possible downclocking on current devices
+const size_t AVX512F_THRESHOLD_SIZE = 128 * 1024;
```
It's based on Dan's numbers for both the downclocking impact (perhaps 30% over 700 microseconds or ~200 microseconds), and the speed advantage for AVX512F over AVX2 (about 400 microseconds at 128*1024). Dan's speed numbers are for relatively low clock rate machines (2.8 and 2.1 GHz), so I think on current hardware, the time difference may be lower due to higher clock rates on both.

I think it's just a rough estimate, but not very wrong.

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -22032,7 +22306,40 @@ void gc_heap::plan_phase (int condemned_gen_number)
+        {
+            int32_t* mark_list_32 = (int32_t*)mark_list;
+            uint8_t* low = gc_low;
+            ptrdiff_t range = heap_segment_allocated (ephemeral_heap_segment) - low;
```
Ah, I see, I wasn't aware of these - thanks!

Maoni0 replied:
sorry, I should've been more clear - shigh is actually a valid object that we marked, just the highest one we marked. so the range should be `shigh + Align (size (shigh)) - slow`

PeterSolMS commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -8276,9 +8347,142 @@ void gc_heap::sort_mark_list()
+    }
+
+    // give up if this is not an ephemeral GC or the mark list size is unreasonably large
+    if (settings.condemned_generation > 1 || total_mark_list_size > total_ephemeral_size/256)
```
Ok, moved the test for eph GC up and changed it to compare against max_generation.

256 is a number based on a rough back-of-the envelope computation. The idea was that walking the heap will be dominated by fetching cache lines, and for each marked object we will roughly fetch 2 cache lines (one for the marked object, and one for the unmarked object after it). Cache line size is 64 bytes, so if number of mark list elements is ephemeral_size/256, we will fetch about half the cache lines. So at this threshold, there should still be a bit of benefit for using the mark list, but probably not a whole lot, because just walking the heap would benefit from prefetching etc.

So that was my reasoning...

PeterSolMS committed:
Address code review feedback:
 - fix typo in comment in InitSupportedInstructionSet
 - move test for full GC to beginning of sort_mark_list
 - in WKS GC, we can use the tighter range shigh - slow for the surviving objects instead of the full ephemeral range.
 - make the description for the new config setting GCEnabledInstructionSets more explicit by enumerating the legal values and their meanings.

PeterSolMS committed:
Snapshot for Linux changes

PeterSolMS committed:
Merge remote-tracking branch 'origin/vxsort' into vxsort

PeterSolMS committed:
Add more definitions to immintrinh.h

PeterSolMS committed:
Fix cmake warnings about mismatched endif clauses.

PeterSolMS committed:
Disable Linux support for now due to multiple compile & link errors.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -8276,9 +8352,142 @@ void gc_heap::sort_mark_list()
+        {
+            dprintf (3, ("Sorting mark lists as 32-bit offsets"));
+
+//#define WRITE_SORT_DATA
```
seems weird to have a #define in the middle of its usage...

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -8276,9 +8352,142 @@ void gc_heap::sort_mark_list()
+                int log2_item_count = index_of_highest_set_bit (item_count);
+                double elapsed_cyles_by_n_log_n = (double)elapsed_cycles / item_count / log2_item_count;
+
+//                printf ("GC#%d: first phase of sort_mark_list for heap %d took %u cycles to sort %u entries (cost/(n*log2(n) = %5.2f)\n", settings.gc_index, this->heap_number, elapsed_cycles, item_count, elapsed_cyles_by_n_log_n);
```
are you actually using the printf's in this method? if not we should just get rid of them... or convert to dprintf's...

PeterSolMS replied:
I converted the printfs to dprintfs.

The call to introsort is a debug only check. We can get rid of it once we fully trust the vectorized sort.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -8276,9 +8352,142 @@ void gc_heap::sort_mark_list()
+    }
+
+    // give up if this is not an ephemeral GC or the mark list size is unreasonably large
+    if (total_mark_list_size > total_ephemeral_size/256)
```
nit - styling 

`if (total_mark_list_size > (total_ephemeral_size / 256))`

PeterSolMS replied:
Ok, done.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -22032,7 +22311,40 @@ void gc_heap::plan_phase (int condemned_gen_number)
+        ptrdiff_t entry_count = mark_list_index - mark_list;
+        // conservatively use AVX2 only for large mark lists,
+        // and do runtime test to check whether AVX2 is indeed available 
+        if (entry_count > AVX2_THRESHOLD_SIZE && IsSupportedInstructionSet (InstructionSet::AVX2))
```
nit -
```cplusplus
if ((entry_count > AVX2_THRESHOLD_SIZE) && IsSupportedInstructionSet (InstructionSet::AVX2))
```

PeterSolMS replied:
Ok, done.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -22032,7 +22311,40 @@ void gc_heap::plan_phase (int condemned_gen_number)
+                for (ptrdiff_t i = 0; i < entry_count; i++)
+                {
+                    uint8_t* item = mark_list[i];
+                    assert (low <= item && item < high);
```
you could changed this assert to

`assert ((low <= item) && (item <= shigh));`

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -22032,7 +22311,40 @@ void gc_heap::plan_phase (int condemned_gen_number)
+            ptrdiff_t range = shigh - low;
+            if ((uint32_t)range == range)
+            {
+                do_pack_avx2 (mark_list, entry_count, low);
+                _sort (&mark_list_32[0], &mark_list_32[entry_count - 1], 0);
+                do_unpack_avx2 (mark_list_32, entry_count, low);
```
can we get the pack/int32sort/unpack in one method, something like try_vxsort_32 so we can call it both in `plan_phase` and `sort_mark_list`?

PeterSolMS replied:
That's a good idea - in fact I moved the code that checks the range, the instruction set etc. into one method do_vxsort.

Maoni0 commented on src/coreclr/src/gc/gcsvr.cpp:
```diff
@@ -21,6 +21,10 @@
 
 #define SERVER_GC 1
 
+#if defined(TARGET_AMD64)
```
I suppose you might as well also change this to check for `TARGET_WINDOWS` to be consistent.

PeterSolMS replied:
Right, that slipped throught the cracks - thanks!

Maoni0 commented on src/coreclr/src/gc/gcwks.cpp:
```diff
@@ -21,6 +21,10 @@
 #undef SERVER_GC
 #endif
 
+#if defined(TARGET_AMD64)
```
I suppose you might as well also change this to check for `TARGET_WINDOWS` to be consistent.

PeterSolMS replied:
Yep - thanks!

Maoni0 commented:
can we add some instructions in the bitonic sort codegen dir for how to run the python scripts there if we need to change the length for which we want to sort with bitonic sort? I doubt we'll be doing that any time soon but it's always good to know how if we do want to experiment with it.

PeterSolMS commented:
I added a comment at the top of vxsort/smallsort/bitonic_gen.py and verified that running it with the suggested arguments actually does produce the code we have.

PeterSolMS committed:
Address code review feedback:
 - add instructions to bitonic_gen.py
 - centralize range and instruction set checks in do_vxsort
 - add parentheses around expressions.
 - removed some printfs, converted others to dprintf
 - strengthened assert

Maoni0 commented:
all the failures are due to eventpipe test failing which is tracked by [#39361](https://github.com/dotnet/runtime/issues/39361)

stephentoub commented:
> measuring and analyzing perf impact is a big part of what Peter has been doing; and will be included as part of the commit message when we merge.

Did this happen?

Maoni0 commented:
it happened and I've chatted with Peter and there's more to do on our side. but we wanted to get this in for Preview8 so folks can try it and help flush out bugs if any. in addition to testing on our side, we had one of our 1st party customers try it and they saw 10% ephemeral pause reduction. obviously as with any perf change this is not a universal number (we had another 1st party customer try this and they saw no change as they already didn't spend much time sorting the mark list). 

in .NET 6 I'd like to make it so that we always include perf results for perf changes in GC.
